### PR TITLE
[BEAM-86] Add CountingInput as a PTransform

### PR DESCRIPTION
**Pull request by tgroh**
This transform produces an unbounded PCollection containing longs based
on a CountingSource.

Deprecate methods producing a Source in CountingSource.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @dhalperi

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingInput.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import com.google.cloud.dataflow.sdk.values.PCollection;
+
+import org.joda.time.Instant;
+
+public class CountingInput {
+  /**
+   * Creates a {@link BoundedCountingInput} that will produce the specified number of elements,
+   * from {@code 0} to {@code numElements - 1}.
+   */
+  public static PTransform<PBegin, PCollection<Long>> upTo(long numElements) {
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingInput.java]: return type -- BoundedCountingInput?

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingInput.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+   * <p>After {@link Long#MAX_VALUE}, the transform never produces more output. (In practice, this
+   * limit should never be reached.)
+   *
+   * <p>Elements in the resulting {@link PCollection PCollection&lt;Long&gt;} will by default have
+   * timestamps corresponding to processing time at element generation, provided by
+   * {@link Instant#now}. Use the transform returned by
+   * {@link UnboundedCountingInput#withTimestampFn(SerializableFunction)} to control the output
+   * timestamps.
+   */
+  public static PTransform<PBegin, PCollection<Long>> unbounded() {
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingInput.java]: return type?

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingSource.java [hunk 3]
```diff
@@ -52,16 +53,19 @@
  * PCollection<Long> bounded = p.apply(Read.from(source));
  * }</pre>
  *
- * <p>To produce an unbounded {@code PCollection<Long>}, use {@link CountingSource#unbounded} or
- * {@link CountingSource#unboundedWithTimestampFn}:
+ * <p>To produce an unbounded {@code PCollection<Long>}, use
+ * {@link Countinginput#unbounded()}, calling
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingSource.java]: hard to tell, but maybe wrapping?

Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingSource.java]: CountingInput

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingSource.java [hunk 4]
```diff
@@ -52,16 +53,19 @@
  * PCollection<Long> bounded = p.apply(Read.from(source));
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingSource.java]: fix up here, too?

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingSource.java [hunk 5]
```diff
@@ -52,16 +53,19 @@
...
  *
  * // To create an unbounded source that uses processing time as the element timestamp.
- * UnboundedSource<Long, CounterMark> source = CountingSource.unbounded();
+ * PTransform<PBegin, PCollection<Long>> producer = CountingInput.unbounded();
  * // Or, to create an unbounded source that uses a provided function to set the element timestamp.
- * UnboundedSource<Long, CounterMark> source = CountingSource.unboundedWithTimestampFn(someFn);
+ * PTransform<PBegin, PCollection<Long>> producer =
+ *     CountingInput.unbounded().withTimestampFn(someFn);
  *
  * PCollection<Long> unbounded = p.apply(Read.from(source));
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingSource.java]: producer? not source.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingInput.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  /**
+   * Creates a {@link BoundedCountingInput} that will produce the specified number of elements,
+   * from {@code 0} to {@code numElements - 1}.
+   */
+  public static PTransform<PBegin, PCollection<Long>> upTo(long numElements) {
+    checkArgument(numElements > 0, "numElements (%s) must be greater than 0", numElements);
+    return new BoundedCountingInput(numElements);
+  }
+
+  /**
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingInput.java]: need tests for this file.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingInput.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+import com.google.cloud.dataflow.sdk.io.CountingSource.NowTimestampFn;
+import com.google.cloud.dataflow.sdk.transforms.PTransform;
+import com.google.cloud.dataflow.sdk.transforms.SerializableFunction;
+import com.google.cloud.dataflow.sdk.values.PBegin;
+import com.google.cloud.dataflow.sdk.values.PCollection;
+
+import org.joda.time.Instant;
+
+public class CountingInput {
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingInput.java]: class-level javadoc mimicking the CountingSource javadoc.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingInput.java (hunk 1 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingInput.java]: Done.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingInput.java (hunk 6 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingInput.java]: Done

**Comment by tgroh** [comment]
I haven't added a checkNotNull to the constructor, as it's private and all of the withers do the assertion directly. Can still add if required.

Other than that, done.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingInput.java [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+ * with timestamps other than {@link Instant#now}.
+ *
+ * <pre>{@code
+ * Pipeline p = ...
+ *
+ * // To create an unbounded producer that uses processing time as the element timestamp.
+ * PTransform<PBegin, PCollection<Long>> producer = CountingInput.unbounded();
+ * // Or, to create an unbounded source that uses a provided function to set the element timestamp.
+ * PTransform<PBegin, PCollection<Long>> producer =
+ *     CountingInput.unbounded().withTimestampFn(someFn);
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingInput.java]: mirror the javadoc improvements here?

**Comment by dhalperi** [comment]
LGTM

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingInput.java (hunk 8 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/CountingInput.java]: Done

**Comment by tgroh** [comment]
Squashed together all of the commits.

**Comment by tgroh** [comment]
Closed by 5a7bd8

**Pull request closed without merging**